Equation of state and bulk viscosity damping as extra stress, closes #9548.

ngrilli opened the pull request:
In classes derived from ComputeExtraStressBase, only _stress_old at the previous time step can be used, but not the _stress at the current time step, because it leads to a cyclic dependency of the material properties.

The abstract class "ComputeStressEosBase" subtracts the volumetric part of the current _stress, calculated by a constitutive model, and adds an equation of state and bulk viscosity terms, useful in shock physics.

closes #9548

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/103137/) on 9bdfd67 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format f83040c3976ab38df6f39f3c80a694678bb44a06`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9551/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9551/style.patch | git apply -v`

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/103223/) on 6bf6e8b wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9551/site)

jwpeterson commented:
Hi @ngrilli this PR looks like it needs to have regression tests with it. Are you working with any of the other MOOSE developers for this PR?

ngrilli commented:
No, I am not sure how to make tests for an abstract class like this one. Should I write an example class derived from ComputeStressEosBase and then write a test for that?

jwpeterson commented:
> Should I write an example class derived from ComputeStressEosBase and then write a test for that?

Yes, otherwise the new code will negatively impact our code coverage stats. We are currently working on a more formalized way of allowing people to add test classes in the modules, but for the time being you can create include/tests and src/tests directories and place your test object there. That way @brianmoose will know what to move when he updates all the modules.

jwpeterson commented:
While you are adding the new test, go ahead and squash your two style change commits into the first commit to clean up the history a bit.

sapitts commented on modules/tensor_mechanics/include/materials/ComputeStressEosBase.h:
```diff
@@ -0,0 +1,47 @@
+ * that is substituted to the volumetric stress
+ * calculated by the constitutive model.
+ */
+class ComputeStressEosBase;
```
I'd prefer to have `ComputeEquationOfStateStressBase` rather than `Eos` in the class name and in the class description.  The term equation of state also seems too broad for your implementation of Birch-Murnaghan, so perhaps `ComputeBirchMurnaghanEquationOfStress` is better, without the base class designation.

ngrilli replied:
Name has been changed to ComputeBirchMurnaghanEquationOfStress

sapitts commented on modules/tensor_mechanics/src/materials/ComputeStressEosBase.C:
```diff
@@ -0,0 +1,70 @@
+  bulk_viscosity_stress.addIa(_C1 * dspecific_volume_dt);
+
+  // Subtract the original volumetric stress, add the one given by the EOS and bulk viscosity
+  _stress[_qp] = _stress[_qp] + EOS_stress + bulk_viscosity_stress - volumetric_stress;
```
use += here for clarity

ngrilli replied:
+= used

sapitts commented on modules/tensor_mechanics/src/materials/ComputeStressEosBase.C:
```diff
@@ -0,0 +1,70 @@
+
+  bulk_viscosity_stress.zero();
+  // Calculate bulk viscosity damping
+  bulk_viscosity_stress.addIa(_C0 * dspecific_volume_dt * abs(dspecific_volume_dt));
```
use `std::abs`

ngrilli replied:
std:abs used

sapitts commented on modules/tensor_mechanics/include/materials/ComputeStressEosBase.h:
```diff
@@ -0,0 +1,47 @@
+#include "ComputeStressBase.h"
+
+/**
+ * ComputeStressEosBase adds a volumetric extra stress including a Birch-Murnaghan EOS
```
Include the citation information if you are implementing equations from a specific paper (e.g. Birch Physical Review 71 (1947) 809-824)

ngrilli replied:
Citations have been included

sapitts reviewed (changes requested):
This class appears too specialized to be a base class, and better documentation in the description is needed.  Perhaps the addition of your testing class and tests will better clarify your design choice in using the base class designation here.  Be careful ( and avoid as much as possible) with your use of abbreviations which can confuse people outside of your field.

Are you also planning to add a strain calculator as well?  If yes, I'd like to see that strain calculator in this commit as well for a complete functionality set.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/104270/) on 954e38c wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format e9f31a04fd5c0fa234e695c9c3aefe5dfe5a3ca0`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9551/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9551/style.patch | git apply -v`

ngrilli commented:
I do not plan to add a strain calculator. ComputeFiniteStrain is sufficient to calculate the deformation gradient that is needed for the equation of state and bulk viscosity term

sapitts commented on modules/tensor_mechanics/src/materials/ComputeBirchMurnaghanEquationOfStress.C:
```diff
@@ -0,0 +1,85 @@
+  // Calculate bulk viscosity damping
+  // C0 * dot(J) / J * |dot(J) / J| + C1 * dot(J) / J
+  // C0 should be chosen of the order of rho Le^2, rho = density, Le = element size
+  // C1 should be chosen of the order of rho Le cs, cs = sound speed
```
Its unclear what you mean by the `rho Le` type comments--are these two quantities multiplied by each other?

ngrilli replied:
It is a multiplication, I have added the * symbol

sapitts commented on modules/tensor_mechanics/src/materials/ComputeBirchMurnaghanEquationOfStress.C:
```diff
@@ -0,0 +1,85 @@
+  // P = (K0 / n) * J^{-n}
+  // J is the Jacobian: J = detF = v / v0, F = deformation gradient
+  // v and v0 are the specific volumes: deformed and reference
+  // Austin et al. JOURNAL OF APPLIED PHYSICS 117, 185902 (2015)
```
Will these two blocks of comments with the equations and citations show in the doxygen rendering even though the method is marked override?

ngrilli replied:
I have tried making local documentation by running "doxygen ./doc/doxygen/Doxyfile" from the tensor_mechanics folder. In the resulting html documentation I can see the comments.

sapitts commented on modules/tensor_mechanics/tests/finite_strain_elastic/equation_of_state.i:
```diff
@@ -0,0 +1,135 @@
+# expected stress_zz is: P = (K0 / n) * J^{-n}
+# J is the Jacobian: J = detF = v / v0, F = deformation gradient
+# v and v0 are the specific volumes: deformed and reference
+# Austin et al. JOURNAL OF APPLIED PHYSICS 117, 185902 (2015)
```
Do you have an expected analytical solution for your test case?  If yes, please add the expected solution and the Moose deviation from the analytical as a comment in the input file

ngrilli replied:
I have reported the analytical solution for the component stress_zz, which is equal to the pressure P, given by the equation of state. The value at the final timestep is equal to the analytical value, I have reported this value in the comment

sapitts reviewed (changes requested):
Thank you for adding in the citation information for your code.  Just a couple of more clean up items that I'm requesting changes for, and then this looks good, unless @aeslaughter thinks that the new documentation system is ready for to have a markdown documentation page written for this code.

sapitts reviewed (approved):
Looks good to me--thanks for the changes!

Commit:
New class ComputeBirchMurnaghanEquationOfStress
introduced to add an equation of state and bulk viscosity terms
to the volumetric part of the stress tensor.
Useful in shock physics.

Added test class ComputeFiniteStrainElasticStressBirchMurnaghan
and test input file equation_of_state.i

Added citations:
Austin et al. JOURNAL OF APPLIED PHYSICS 117, 185902 (2015)
Maheo et al. Mechanics Research Communications 38 (2011) 81 88

closes #9548

jwpeterson reviewed (changes requested):
Your commit message seems to have gotten lost/greatly truncated during the rebasing process (I recall it being fine before?) If you can fix that up, I'll get this merged. Some guidelines on writing good commit messages can be found [here](https://chris.beams.io/posts/git-commit).
